tests/fixture/tmpnet/monitor_processes.go

```diff
@@ -0,0 +1,496 @@
+		return 0, fmt.Errorf("failed to read %s PID file %s: %w", cmdName, pidPath, err)
+	}
+	if len(pidData) == 0 {
+		return 0, nil
+	}
+	pid, err := strconv.Atoi(string(pidData))
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse %s PID: %w", cmdName, err)
+	}
+	return pid, nil
+}
+
+// clearStalePIDFile remove an existing pid file to avoid conflicting with a new process.
+func clearStalePIDFile(log logging.Logger, cmdName string, pidPath string) error {
+	if err := os.Remove(pidPath); err != nil {
+		if !errors.Is(err, fs.ErrNotExist) {
+			return fmt.Errorf("failed to remove stale pid file: %w", err)
+		}
+	} else {
+		log.Info("deleted stale "+cmdName+" pid file",
+			zap.String("path", pidPath),
+		)
+	}
+	return nil
+}
+
+func getPrometheusURL() string {
+	return GetEnvWithDefault("PROMETHEUS_URL", defaultPrometheusURL)
+}
+
+func getLokiURL() string {
+	return GetEnvWithDefault("LOKI_URL", defaultLokiURL)
+}
+
+// getCollectorCredentials retrieves the username and password for the command.
+func getCollectorCredentials(cmdName string) (string, string, error) {
+	var baseEnvName string
+	switch cmdName {
+	case prometheusCmd:
+		baseEnvName = "PROMETHEUS"
+	case promtailCmd:
+		baseEnvName = "LOKI"
+	default:
+		return "", "", fmt.Errorf("unsupported cmd: %s", cmdName)
+	}
+
+	usernameEnvVar := baseEnvName + "_USERNAME"
+	username := GetEnvWithDefault(usernameEnvVar, "")
+	if len(username) == 0 {
+		return "", "", fmt.Errorf("%s env var not set", usernameEnvVar)
+	}
+	passwordEnvVar := baseEnvName + "_PASSWORD"
+	password := GetEnvWithDefault(passwordEnvVar, "")
+	if len(password) == 0 {
+		return "", "", fmt.Errorf("%s var not set", passwordEnvVar)
+	}
+	return username, password, nil
+}
+
+// Start a collector process. Use bash to execute the command in the background and enable
+// stderr and stdout redirection to a log file.
+//
+// Ideally this would be possible without bash, but it does not seem possible to
+// have this process open a log file, set cmd.Stdout cmd.Stderr to that file, and
+// then have the child process be able to write to that file once the parent
+// process exits. Attempting to do so resulted in an empty log file.
+func startCollectorProcess(
+	ctx context.Context,
+	log logging.Logger,
+	cmdName string,
+	args string,
+	workingDir string,
+	pidPath string,
+) error {
+	logFilename := cmdName + ".log"
+	fullCmd := "nohup " + cmdName + " " + args + " > " + logFilename + " 2>&1 & echo -n \"$!\" > " + pidPath
+	log.Info("starting "+cmdName,
+		zap.String("workingDir", workingDir),
+		zap.String("fullCmd", fullCmd),
+	)
+
+	cmd := exec.Command("bash", "-c", fullCmd)
+	configureDetachedProcess(cmd) // Ensure the child process will outlive its parent
+	cmd.Dir = workingDir
+	if err := cmd.Start(); err != nil {
+		return fmt.Errorf("failed to start %s: %w", cmdName, err)
+	}
+
+	// Wait for PID file
+	var pid int
+	if err := pollUntilContextCancel(ctx, func(_ context.Context) (bool, error) {
```

Comment:
Small style nit: I find the indentation here pretty odd. Specifically because the `return pid != 0, nil` happens at the same indentation level as `return err`, but they are actually referring to different functions entirely.

Personally, I try to avoid multi-line `if statement; condition {`s because it just looks really weird to me.